multi auto link on operator drop

MysteriousChallenger opened the pull request:
When dropping an operator onto the graph, automatically connect to all possible inputs and outputs.

DaiNiaoZhou commented:
The design looks good to me except that I think a binary search is an overkill.

MysteriousChallenger commented:
Added Jasmine tests. 

Minor changes from last iteration: 
- onDrop now proactively searches for suggested ops. Previously took results from latest search (called by onDrag event)
- Added mock operator schema and mock graph data for a dummy op with 3 inputs / 3 outputs

MysteriousChallenger commented:
Changes since last iteration:

- Amended Jasmine tests to test multiple links being created at once
- onDrop now triggers search for suggested operators/links. Previously took results from last search (called by onDrag event)
- Added mock operator schema and mock workflow data for a dummy operator with 3 inputs and 3 outputs

MysteriousChallenger commented:
Has this branch been merged properly? It still says "4 commits ahead of master." Sorry, a little new to github.

DaiNiaoZhou commented:
> Has this branch been merged properly? It still says "4 commits ahead of master." Sorry, a little new to github.

Not yet, Zuozhi needs to approve it for the 4 new commits you have created to merge into the master branch. Your branch is "ahead of the master" means there are commits that exists on your branch but not on the master, vice versa.

zuozhiw commented on core/new-gui/src/app/workspace/service/drag-drop/drag-drop.service.ts:
```diff
@@ -334,105 +342,174 @@ export class DragDropService {
+          && distanceFromCurrentOperator < inputMinDists[numInputOps - 1] && hasFreeOutputPorts(operator)) {
+          const operatorIndex = sortedIndex(inputMinDists, distanceFromCurrentOperator);
+          // splice in new min-dist and new op
+          inputMinDists.splice(operatorIndex, 0, distanceFromCurrentOperator);
```
These lines form the main recommendation mechanism. This mechanism works very well and feels natural to the users. 
However, I feel how this mechanism works is not well documented. The logic might look obvious to you, but it took me quite some time to understand it. I believe it is important to let other people easily understand a piece of code, as the people who maintain the code in the future might not be the original author.  

From what I understand, you are maintaining a sorted array of top n operators (n being number of input/outputs) that are the closest to the current operator. It would be nice if you are add such kind of high level description.
Moreover, I feel it is most natural to use a priority queue for this purpose. Rather than hand-writing the primitive functions like inserting, deleting, and finding sorted index, I prefer to use a library, which makes the code shorter and eaiser to understand. You can try https://github.com/mourner/tinyqueue
The library allows you to store object and use custom comparator. So I would create a priority queue of tuple `[OperatorPredicate, number]` and pass a comparator. In this way, this function would be shorter and easier to read.

MysteriousChallenger replied:
On review, the improvements probably weren't worth it. Too small compared to the "performance concern" functions. I will be refactoring this.

zuozhiw commented:
In general this PR is very good. I have left some more comments, but it's mostly related to coding styles.

I found a bug when testing this feature, when dragging an operator just outside of the threshold range, the highlighting disappears, but when dropping it, the link is still added. Please investigate this bug. Thank you.
![highlight_bug](https://user-images.githubusercontent.com/12578068/79064621-551e2e00-7c5f-11ea-9bf3-3a7f4b1e95e1.gif)

MysteriousChallenger commented on core/new-gui/src/app/workspace/service/drag-drop/drag-drop.service.ts:
```diff
@@ -109,33 +105,49 @@ export class DragDropService {
+        const newLinks: OperatorLink[] = [];
+        const graph = this.workflowActionService.getJointGraphWrapper();
+
+        // search for nearby operators as suggested I/O operators
```
Good to know. will fix that in next commit

MysteriousChallenger commented on core/new-gui/src/app/workspace/service/drag-drop/drag-drop.service.ts:
```diff
@@ -109,33 +105,49 @@ export class DragDropService {
+          });
         }
+        this.workflowActionService.addOperatorsAndLinks(operatorsAndPositions, newLinks);
+        this.resetSuggestions();
```
Agreed. Shortened in new commit

MysteriousChallenger commented on core/new-gui/src/app/workspace/service/drag-drop/drag-drop.service.ts:
```diff
@@ -334,105 +342,174 @@ export class DragDropService {
+      return low;
+    };
+    // TODO: these two functions are a huge performance concern: these operations are expensive and cause lag when dragging
+    const hasFreeOutputPorts = function(operator: OperatorPredicate): boolean {
```
Will do

MysteriousChallenger commented on core/new-gui/src/app/workspace/service/drag-drop/drag-drop.service.ts:
```diff
@@ -334,105 +342,174 @@ export class DragDropService {
+    // remove trailing null vals if not enough inputs/outputs found to fill all ports
+    strip_null(inputOps);
+    strip_null(outputOps);
+    return [<OperatorPredicate[]>inputOps, <OperatorPredicate[]>outputOps];
```
This has been fixed by removing nulls from the process entirely

MysteriousChallenger commented on core/new-gui/src/app/workspace/service/drag-drop/drag-drop.service.ts:
```diff
@@ -334,105 +342,174 @@ export class DragDropService {
+      }
+      return low;
+    };
+    // TODO: these two functions are a huge performance concern: these operations are expensive and cause lag when dragging
```
> 
> 
> In general this PR is very good. I have left some more comments, but it's mostly related to coding styles.
> 
> I found a bug when testing this feature, when dragging an operator just outside of the threshold range, the highlighting disappears, but when dropping it, the link is still added. Please investigate this bug. Thank you.
> ![highlight_bug](https://user-images.githubusercontent.com/12578068/79064621-551e2e00-7c5f-11ea-9bf3-3a7f4b1e95e1.gif)

Turns out an "improvement" I made a few commits ago caused this: upon dropping the floating operator, we would recalculate the nearby operators and connect to those. The logic was that the last set of mouse coordinates would yield the most accurate results. However this is sometimes inconsistent with the operators highlighted by the mouseMove event handler (still not totally sure why), even though the selection criteria is the same. I reverted to using the last calculated nearby operators (calculated from the last mouseMove). The highlighted operators are now always the ones that the new operator connects to.

MysteriousChallenger commented:
> 
> 
> In general this PR is very good. I have left some more comments, but it's mostly related to coding styles.
> 
> I found a bug when testing this feature, when dragging an operator just outside of the threshold range, the highlighting disappears, but when dropping it, the link is still added. Please investigate this bug. Thank you.
> ![highlight_bug](https://user-images.githubusercontent.com/12578068/79064621-551e2e00-7c5f-11ea-9bf3-3a7f4b1e95e1.gif)

Turns out an "improvement" I made a few commits ago caused this: upon dropping the floating operator, we would recalculate the nearby operators and connect to those. The logic was that the last set of mouse coordinates would yield the most accurate results. However this is sometimes inconsistent with the operators highlighted by the mouseMove event handler (still not totally sure why), even though the selection criteria is the same. I reverted to using the last calculated nearby operators (calculated from the last mouseMove). The highlighted operators are now always the ones that the new operator connects to.

zuozhiw commented:
@MysteriousChallenger the changes look good. When I'm testing it, I found another bug, please fix it.
There's also a place where you added `ts-ignore`, I understand the reasoning, but I prefer not to add `ts-ignore`. In order to work around the undefined check, please use something like:
```
const top = inputOps.peek();
if (top) {
// more code
}
```

![Untitled](https://user-images.githubusercontent.com/12578068/79547182-680e7500-8048-11ea-85aa-57ac2b6d476c.gif)

MysteriousChallenger commented:
> 
> 
> @MysteriousChallenger the changes look good. When I'm testing it, I found another bug, please fix it.
> There's also a place where you added `ts-ignore`, I understand the reasoning, but I prefer not to add `ts-ignore`. In order to work around the undefined check, please use something like:
> 
> ```
> const top = inputOps.peek();
> if (top) {
> // more code
> }
> ```
> 
> ![Untitled](https://user-images.githubusercontent.com/12578068/79547182-680e7500-8048-11ea-85aa-57ac2b6d476c.gif)

That was embarrassing: I mixed some variable names and even misplaced an if-else block. It is fixed for sure this time, so help me god. I ran unit tests and tried it myself before this commit.

sorry for taking up more time
